lmformatenforcer/tokenizerprefixtree.py

```diff
@@ -8,46 +8,95 @@ def __init__(self) -> None:
+        token_lists = {}
+        max_token_len = max(len(token) for token in all_tokens)
+        for min_remaining in range(max_token_len + 1):
+            for max_len in range(max_token_len + 1):
+                if max_len >= min_remaining:  # Skip combinations that are never used
+                    token_lists[(min_remaining, max_len)] = tuple(sorted([
+                        token for token in all_tokens
+                        if _valid_for_min_remaining(token, min_remaining) and _valid_for_max_len(token, max_len)
+                    ]))
+
+        # Deduplicate the lists to save RAM as many of them will be identical
+        unique_lists = set(token_lists.values())
+        for key, lst in token_lists.items():
+            for uniq in unique_lists:
+                if len(uniq) == len(lst) and uniq == lst:
+                    token_lists[key] = uniq
+                    break
+
+        # Turn token strings into token numbers
+        self.allowlist_cache = {
+            key: tuple(self.token_str_to_num[t] for t in lst)
+            for key, lst in token_lists.items()
+        }
+        del self.token_str_to_num
+
+
 class TokenizerPrefixTree:
     def __init__(self, regular_tokens: List[Tuple[int, str, bool]]):
         self.root = TokenizerPrefixTreeNode()
-        self.json_freetext_tokens: OrderedDict[int, List[int]] = OrderedDict()  # { max_token_len: [token_ids] }
+        self.json_freetext_tokens = JsonFreetextTokenCache()
         self.new_word_tokens: Set[int] = set()
         self.tokens_to_strs = {token_idx: token_str for token_idx, token_str, _ in regular_tokens}
         for token_idx, decoded, is_new_word in regular_tokens:
             self._add_token_to_tree(decoded, token_idx, self.root)
+
             # Performance optimization - cache the tokens of all the strings that don't contain a quote in the middle, or a line break.
             # When we are in a JSON freetext string field, they will all be permitted and this will save a lot of tree iterations.
+            has_non_trailing_backslash = '\\' in decoded[:-1]
             has_quote_before_end = '"' in decoded[0:-1]
-            has_newline = "\n" in decoded or "\r" in decoded
-
-            if not (has_quote_before_end or has_newline):
-                if '\\' in decoded[:-1]:
-                    # If there is a backslash that is not trailing, we might be in an illegal json territory. Need to verify
-                    # that is is a legal json character streak
-                    try:
-                        json.loads(f'"{decoded}"')
-                    except json.decoder.JSONDecodeError:
-                        continue
+            if has_non_trailing_backslash or has_quote_before_end:  # Could be illegal inside a JSON string, test it
+                try:
+                    json.loads(f'"{decoded}"')
+                except json.decoder.JSONDecodeError:
+                    continue
```

Comment:
This `continue` short-circuits...
```py
            if is_new_word:
                self.new_word_tokens.add(token_idx)
```
...a few lines later. Is this your intention @noamgat ?